iceoryx_posh/include/iceoryx_posh/internal/popo/sample.inl

```diff
@@ -22,96 +23,27 @@ namespace iox
-{
-template <typename T, typename H>
-inline SamplePrivateData<T, H>::SamplePrivateData(cxx::unique_ptr<T>&& sampleUniquePtr,
-                                                  PublisherInterface<T, H>& publisher) noexcept
-    : sampleUniquePtr(std::move(sampleUniquePtr))
-    , publisherRef(publisher)
-{
-}
-
-template <typename T, typename H>
-inline SamplePrivateData<const T, H>::SamplePrivateData(cxx::unique_ptr<const T>&& sampleUniquePtr) noexcept
-    : sampleUniquePtr(std::move(sampleUniquePtr))
-{
-}
-} // namespace internal
-
 template <typename T, typename H>
 template <typename S, typename>
 inline Sample<T, H>::Sample(cxx::unique_ptr<T>&& sampleUniquePtr, PublisherInterface<T, H>& publisher) noexcept
-    : m_members({std::move(sampleUniquePtr), publisher})
+    : SmartChunk<PublisherInterface<T, H>, T, H>(std::move(sampleUniquePtr), publisher)
 {
 }
 
 template <typename T, typename H>
 template <typename S, typename>
-inline Sample<T, H>::Sample(cxx::unique_ptr<T>&& sampleUniquePtr) noexcept
-    : m_members(std::move(sampleUniquePtr))
-{
-}
-
-template <typename T, typename H>
-inline T* Sample<T, H>::operator->() noexcept
-{
-    return get();
-}
-
-template <typename T, typename H>
-inline const T* Sample<T, H>::operator->() const noexcept
-{
-    return get();
-}
-
-template <typename T, typename H>
-inline T& Sample<T, H>::operator*() noexcept
-{
-    return *get();
-}
-
-template <typename T, typename H>
-inline const T& Sample<T, H>::operator*() const noexcept
+inline Sample<T, H>::Sample(cxx::unique_ptr<const T>&& sampleUniquePtr) noexcept
+    : SmartChunk<PublisherInterface<T, H>, T, H>(std::move(sampleUniquePtr))
 {
-    return *get();
-}
-
-template <typename T, typename H>
-inline Sample<T, H>::operator bool() const noexcept
-{
-    return get() != nullptr;
-}
-
-template <typename T, typename H>
-inline T* Sample<T, H>::get() noexcept
-{
-    return m_members.sampleUniquePtr.get();
-}
-
-template <typename T, typename H>
-inline const T* Sample<T, H>::get() const noexcept
-{
-    return m_members.sampleUniquePtr.get();
-}
-
-template <typename T, typename H>
-inline typename Sample<T, H>::ConditionalConstChunkHeader_t* Sample<T, H>::getChunkHeader() noexcept
-{
-    return mepoo::ChunkHeader::fromUserPayload(m_members.sampleUniquePtr.get());
-}
-
-template <typename T, typename H>
-inline const mepoo::ChunkHeader* Sample<T, H>::getChunkHeader() const noexcept
-{
-    return mepoo::ChunkHeader::fromUserPayload(m_members.sampleUniquePtr.get());
 }
 
 template <typename T, typename H>
 template <typename R, typename>
 inline R& Sample<T, H>::getUserHeader() noexcept
 {
-    return *static_cast<R*>(mepoo::ChunkHeader::fromUserPayload(m_members.sampleUniquePtr.get())->userHeader());
+    return *static_cast<R*>(mepoo::ChunkHeader::fromUserPayload(
+                                SmartChunk<PublisherInterface<T, H>, T, H>::m_members.smartchunkUniquePtr.get())
+                                ->userHeader());
```

Comment:
using just `m_members.smart...` should also work

Reply:
I have to use Base_t here. Since the SmartChunk is templated, the derived class will not know by itself which type of SmartChunk has to be inherited.